### PR TITLE
[ADD] 14.0 cancel and redraft sale endpoint

### PR DESCRIPTION
**Pull request by PierrickBrun**
@sebastienbeau Can you review ?

### CODE REVIEW

File: shopinvader/services/sale.py [hunk 1]
```diff
@@ -49,6 +51,16 @@ def ask_email_invoice(self, _id):
...
 
+    def cancel(self, _id):
+        order = self._get(_id)
+        self._cancel(order)
+        return self._to_json(order)[0]
+
+    def redraft(self, _id):
+        order = self._get(_id)
+        self._cancel(order, redraft=True)
+        return self._to_json(order)[0]
```
Comment (sebastienbeau) [inline shopinvader/services/sale.py]: You should also update the session in order to set the id of this sale as the cart id

**Review by sebastienbeau — CHANGES_REQUESTED** [review]

File: shopinvader/services/sale.py (hunk 1 above)
Comment (PierrickBrun) [reply shopinvader/services/sale.py]: Thanks, done (also I renamed `redraft` to `reset_to_cart`)

File: shopinvader/services/sale.py [hunk 2]
```diff
@@ -102,6 +125,18 @@ def _get_invoices(self, sale):
         domain_state = invoice_service._get_domain_state()
         return invoices.filtered_domain(domain_state)
 
+    def _cancel(self, sale, reset_to_cart=False):
+        for line in sale.order_line:
+            if line.qty_delivered > 0 or line.qty_invoiced > 0:
```
Comment (rousseldenis) [inline shopinvader/services/sale.py]: You should use float_compare or float_is_zero

Comment (rousseldenis) [reply shopinvader/services/sale.py]: IMHO the cancel behaviour could be conditionnable to something else (we agree this is the default).

So, I would have done a function to help people add their condtion for cancelable sales.

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review. Some remarks

File: shopinvader/services/sale.py (hunk 2 above)
Comment (sebastienbeau) [reply shopinvader/services/sale.py]: @PierrickBrun can you fix it?

**Review by sebastienbeau — APPROVED** [review]

**Comment by sebastienbeau** [comment]
@rousseldenis can you update your review

**Review by rousseldenis — APPROVED** [review]

**Pull request merged**